### PR TITLE
Unity Hub Flatpak

### PR DESCRIPTION
**Pull request by refi64**
What works:

- Unity Hub.
- Downloading the Editor (mostly, see below).
- Running and working with the Editor.
- Editing files in VS Code.

What doesn't work:

- VS Code C# and Unity integration extensions, pending https://github.com/flatpak/flatpak/issues/2675.
- For some reason Editor insists on saving config data into `~/.config/unity3d`, ignoring the XDG directories.
- (**EDIT:** Patched in this version until Unity updates the Hub.) There's a Unity Hub bug where it unconditionally checks the rootfs for space instead of the home or tmp directories, I've submitted it and am waiting for feedback. As a result, you can't download anything greater than the size of the Flatpak rootfs tmpfs.

What's ugly:

- VS Code integration is arguably a bit odd.
- The `sed ...` is needed because otherwise Unity Hub splits its own data files between `~/.var/app/com.unity.UnityHub/data` directory and `~/.local/share`, leading to a lot of different oddities and issues. For this and the `~/.config` issue above, I've again submitted a bug and am waiting for feedback.
- Unity Hub does not have a versioned download URL *at all*.

Largely supersedes https://github.com/flathub/flathub/pull/267. Tried to make as user-friendly as possible, main issue is probably going to be setting the proper VS Code script (I'd kind of like to set this automatically but can't figure out how).

### CODE REVIEW

**Comment by ramcq** [comment]
>     * For some reason Editor insists on saving config data into `~/.config/unity3d`, ignoring the XDG directories.

Can we just use `--persist` for this?

>     * There's a Unity Hub bug where it unconditionally checks the rootfs for space instead of the home or tmp directories, I've submitted it and am waiting for feedback. As a result, you can't download anything greater than the size of the Flatpak rootfs tmpfs.

Woooooooowwwwwwwwww.... Ouch.

> What's ugly:
> 
>     * VS Code integration is arguably a bit odd.

Howso? Is there some way we can pre-seed the configuration or massage it with our wrapper script? How is VS Code usually detected?

>     * The `sed ...` is needed because otherwise Unity Hub splits its own data files between `~/.var/app/com.unity.UnityHub/data` directory and `~/.local/share`, leading to a lot of different oddities and issues. For this and the `~/.config` issue above, I've again submitted a bug and am waiting for feedback.
> 
>     * Unity Hub does not have a versioned download URL _at all_.

That's a bit lame too. Would it be possible to also file an issue and request one?

**Comment by kirbyfan64** [comment]
> Can we just use `--persist` for this?

Unfortunately not, `--persist` doesn't seem to combine well with `--filesystem=home`.

> Woooooooowwwwwwwwww.... Ouch.

Yup, I've actually managed to patch this in the latest version though.

> Howso? Is there some way we can pre-seed the configuration or massage it with our wrapper script? How is VS Code usually detected?

Normally, you literally have to navigate to and select the VS Code binary yourself. I'll take a look at doing this next, though it's worth noting that the configuration format is rather...bizarre, looks a bit like this:

```xml
<pref name="kScriptEditorArgs" type="string">IiQoRmlsZSki</pref>
<pref name="kScriptEditorArgs/app/bin/vscode-editor" type="string">IiQoRmlsZSki</pref>
<pref name="kScriptsDefaultApp" type="string">L2FwcC9iaW4vdnNjb2RlLWVkaXRvcg==</pref>
```

where `kScriptsDefaultApp`'s value is `/app/bin/vscode-editor` base64-encoded, and both the `*Args` keys are just `"$(File)"` base64-encoded.

**Comment by kirbyfan64** [comment]
Oh, and:

> That's a bit lame too. Would it be possible to also file an issue and request one?

[Already been asked in November, nothing's changed yet.](https://forum.unity.com/threads/unity-hub-versioned-urls.586498/) Considering the most recent Unity Hub (1.4.0) was December, I think it's safe to assume that, we won't see this change until, *at minimum*, the next release.

**Comment by kirbyfan64** [comment]
Also, a bit more research into `~/.config/unity3d`: for some reason, Unity Editor saves *some* of its preferences to `~/.config/unity3d`, as well as the cache and logs. The *majority* of the preferences, however, are in the proper XDG_DATA_HOME. No clue why there's a discrepancy here.

**Comment by kirbyfan64** [comment]
Okay, so some updates:

- VS Code integration *mostly* works, pending https://github.com/flathub/com.visualstudio.code/pull/73.
- The *mostly* is because the debugger won't work, which I'm investigating now.
- https://github.com/flatpak/flatpak/commit/cd2142888fc4c199723a0dfca1f15ea8788a5483 unfortunately broke --appimage-extract, so I switched to a custom squashfs-tools fork that skips the AppImage ELF heading (pending a rebase to the more up-to-date squashkit, but ATM it works for this just fine).
- The prefs will now automatically be set to use VS Code by default, and the user won't have to select it or manually pick whether to use the OSS or standard version.

In short, if https://github.com/flathub/com.visualstudio.code/pull/73 is merged and I figure out the debugger issue, then this will finally be ready.

**Comment by kirbyfan64** [comment]
And *it's done*!!

There are quite a few hacks in this Flatpak overall, but this is probably the craziest. The VS Code debugging extensions looks for the Unity PID and connects to it via `56000 + (PID % 1000)`, so I just create a "fake" Unity process at a random PID and use socat to forward the socket to the actual Unity Editor socket.

In addition, mono is required *in addition to* dotnet (?). However, in the process of all this, I *did* realize that since I'm opening VS Code manually, I can run a custom script inside via `flatpak run --command=bash`, which is what I ended up doing. This also means that https://github.com/flathub/com.visualstudio.code/pull/73 doesn't *have* to be merged.

The catch: VS Code *must* be started from within Unity Editor for any of this to work, it simply can't be practically applied globally from within VS Code.

**Comment by ramcq** [comment]
bot, build com.unity.UnityHub

**Comment by kirbyfan64** [comment]
Was installing the appdata to the wrong path, just fixed.

**Comment by ramcq** [comment]
bot, build com.unity.UnityHub

**Comment by ramcq** [comment]
bot, build com.unity.UnityHub

File: vscode.sh [hunk 1]
```diff
@@ -68,3 +68,5 @@ done
 zenity --error --no-wrap --title='Visual Studio Code is not installed' \
   --text="Visual Studio Code is required to edit scripts with the Unity Hub Flatpak, please \
 install it from Flathub."
+gdbus call -e -d org.gnome.Software -o /org/gnome/Software/SearchProvider  \
```
Comment (ramcq) [inline vscode.sh]: You can probably use the GtkApplication actions to do the equivalent of `gnome-software --details=...`. In D-Bus monitor I see this:
```
method call time=1552037382.773304 sender=:1.1262 -> destination=org.gnome.Software serial=5 path=/org/gnome/Software; interface=org.gtk.Actions; member=Activate
   string "details"
   array [
      variant          struct {
            string "com.visualstudio.code"
            string ""
         }
   ]
   array [
      dict entry(
         string "cwd"
         variant             array of bytes "/home/ramcq" + \0
      )
   ]
```

Is it possible to use Zenity in a prompting mode, so you can ask the user whether they would like to install, and then launch Software in that case, or exit 1 if they say no?

File: patch-resources.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/python3
+
+
```
Comment (ramcq) [inline patch-resources.py]: It would be great if this script or commit had a high-level description of what this file is, what needs to be changed, what problem is being solved, etc.

File: apply_extra.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/bash
+
+unappimage *.AppImage
+mv squashfs-root/* .
+rm -rf squashfs-root *.AppImage
+# asar is a binary archive format, so the replacement string *must* be the same length
```
Comment (ramcq) [inline apply_extra.sh]: Can probably move this comment into the script.

File: apply_extra.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# as the original.
+patch-resources resources/app.asar
+# sed -i -f - resources/app.asar <<'EOF'
```
Comment (ramcq) [inline apply_extra.sh]: Delete stale code.

File: apply_extra.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+# EOF
+rm AppRun
+mkdir -p export/share/applications
+sed '/X-AppImage/d;s/Exec=AppRun/Exec=start-unityhub/;s/Icon=unityhub/Icon=com.unity.UnityHub/' \
```
Comment (ramcq) [inline apply_extra.sh]: Producing the .desktop file at install like this means that appstream-compose at build time will not be able to pull the metadata out of the .desktop file into the appdata, so you will need to duplicate icon, category, etc info in the appdata. It's a bit safer / more consistent to just copy the icon and .desktop file into this repo so they are always present in the repo. (Provided the description makes clear this a third-party re-packaging of the Unity Hub and not supported by Unity Technologies, it's OK for us to use the icon/etc to refer to their product.)

Comment (ramcq) [reply apply_extra.sh]: (also I believe the FlatHub-ists prefer desktop-file-edit than sed, but shipping the file as data would alleviate this concern anyway)

Comment (ramcq) [reply apply_extra.sh]: Maybe this isn't a real concern at all... I guess we have to maintain our own appdata file which can easily contain the icon and categories anyway...

File: com.unity.UnityHub.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  <id type="desktop">com.unity.UnityHub.desktop</id>
+  <name>Unity Hub</name>
+  <developer_name>Unity Technologies</developer_name>
+  <summary>Real-time game engine</summary>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>LicenseRef-proprietary</project_license>
+  <url type="homepage">https://unity3d.com/</url>
+  <icon type="remote" height="256" width="256">https://raw.githubusercontent.com/kirbyfan64/flathub/unity/Unity.png</icon>
+  <description>
+    <p>Unity is used to create half of the world’s games. Our real time platform, powered by tools and services, offer incredible possibilities for game developers, and creators across industries and applications.</p>
```
Comment (ramcq) [inline com.unity.UnityHub.appdata.xml]: We should add some disclaimer that this Flatpak wrapper is not supported by or verified by Unity Tech - eg https://flathub.org/apps/details/com.slack.Slack

File: vscode.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+unity_pid=$(pidof Unity)
+unity_port=$((56000 + unity_pid % 1000))
+
+read -r -d '' start_vscode <<'EOF'
+target_pid="$1"
+shift
+
+cp /usr/bin/sleep /tmp/Unity
+
+for (( i=4; i < $target_pid; i++ )); do /usr/bin/true; done
```
Comment (ramcq) [inline vscode.sh]: My goodness... Given that we fetch the real Unity pid, is it actually worthwhile that the we loop like this up to 100 times and change the local fake Unity pid? Can't we just do something more like run fake Unity, check the corresponding socat port is available, if not, kill Unity and try again? So in the happy case (56004 is available) we will not loop at all.

File: vscode.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+/tmp/Unity infinity &
+
+if ! [[ -d /usr/lib/sdk/dotnet || -d /usr/lib/sdk/mono5 ]]; then
+  zenity --warning --no-wrap --title='dotnet and Mono SDKs are required' \
+    --text="The dotnet and mono SDK extensions are required to use the Unity debugger integration,
+please install them from Flathub."
```
Comment (ramcq) [inline vscode.sh]: Maybe also launch Software in this case, so we could factor the zenity/dbus invocation into a function.

File: vscode.sh (hunk 1 above)
Comment (kirbyfan64) [reply vscode.sh]: @ramcq I'll put Zenity in prompting mode. I can change the software call if you want, but one of the advantages of the SearchProvider route is that the user can pick between VS Code proprietary or OSS. Then again, I'm not sure how much that would matter given that Unity Hub itself is proprietary and probably has telemetry in it (some people have an odd double standard with MS though).

Comment (ramcq) [reply vscode.sh]: I got the search view in G-S from the above too, because it's an ambiguous app ID match.

Comment (kirbyfan64) [reply vscode.sh]: Well after like 30 minutes of debugging, turns out [details is broken in Software for Flatpaks at the moment](https://gitlab.gnome.org/GNOME/gnome-software/issues/577).

**Comment by nedrichards** [comment]
bot, build com.unity.UnityHub

**Comment by nedrichards** [comment]
bot, build com.unity.UnityHub

File: com.unity.UnityHub.appdata.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    <content_attribute id="sex-themes">none</content_attribute>
+    <content_attribute id="sex-homosexuality">none</content_attribute>
+    <content_attribute id="sex-prostitution">none</content_attribute>
+    <content_attribute id="sex-adultery">none</content_attribute>
+    <content_attribute id="sex-appearance">none</content_attribute>
+    <content_attribute id="language-profanity">none</content_attribute>
+    <content_attribute id="language-humor">none</content_attribute>
+    <content_attribute id="language-discrimination">none</content_attribute>
+    <content_attribute id="social-chat">none</content_attribute>
+    <content_attribute id="social-info">none</content_attribute>
```
Comment (ramcq) [inline com.unity.UnityHub.appdata.xml]: I would say at least "mild" here because an account is created on Unity's systems.

**Comment by ramcq** [comment]
@nedrichards @barthalion I'm not exactly a reviewer but this submission seems reasonably good to me - any next steps to move this forward? It would be very popular with... my stakeholders. :)

File: com.unity.UnityHub.appdata.xml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    <p>NOTE: This wrapper is not verified by, affiliated with, or supported by Unity Technologies.</p>
+  </description>
+  <screenshots>
+    <screenshot type="default">
+      <image type="source">https://blogs.unity3d.com/wp-content/uploads/2018/09/image1-3.png</image>
+    </screenshot>
+  </screenshots>
+  <kudos>
+    <kudo>HiDpiIcon</kudo>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: I __think__ Unity uses GTK+ 3 now, so `ModernToolkit` too I guess.

File: com.unity.UnityHub.appdata.xml [hunk 11]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<application>
+  <id type="desktop">com.unity.UnityHub.desktop</id>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: Don’t use a `.desktop` suffix, instead use a proper ID and add a `<launchable>` entry linking to the desktop file: https://github.com/flathub/flathub/wiki/AppData-Guidelines#id

File: com.unity.UnityHub.appdata.xml [hunk 12]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<application>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: This is the old scheme; run `appstream-util upgrade` on the thing.

File: com.unity.UnityHub.appdata.xml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    <kudo>HiDpiIcon</kudo>
+  </kudos>
+  <categories>
+    <category>Development</category>
+  </categories>
+  <releases>
+    <release version="1.6.0" date="2019-03-07"/>
+  </releases>
+  <content_rating type="oars-1.1">
+    <content_attribute id="violence-cartoon">none</content_attribute>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: The `none` values here aren’t wrong, but you can leave them out to increase readability if you want.

File: com.unity.UnityHub.appdata.xml [hunk 14]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<application>
+  <id type="desktop">com.unity.UnityHub.desktop</id>
+  <name>Unity Hub</name>
+  <developer_name>Unity Technologies</developer_name>
+  <summary>Real-time game engine</summary>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>LicenseRef-proprietary</project_license>
+  <url type="homepage">https://unity3d.com/</url>
+  <icon type="remote" height="256" width="256">https://raw.githubusercontent.com/kirbyfan64/flathub/unity/Unity.png</icon>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: This will be pulled in from the desktop file, don’t override it here.

File: com.unity.UnityHub.appdata.xml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  </description>
+  <screenshots>
+    <screenshot type="default">
+      <image type="source">https://blogs.unity3d.com/wp-content/uploads/2018/09/image1-3.png</image>
+    </screenshot>
+  </screenshots>
+  <kudos>
+    <kudo>HiDpiIcon</kudo>
+  </kudos>
+  <categories>
```
Comment (jurf) [inline com.unity.UnityHub.appdata.xml]: Categories are pulled from the desktop file too, so the same thing as with the icon.

File: com.unity.UnityHub.yaml [hunk 16]
```diff
@@ -0,0 +1,84 @@
+app-id: com.unity.UnityHub
+branch: stable
```
Comment (jurf) [inline com.unity.UnityHub.yaml]: `branch` is implied stable on Flathub, you can remove this, then it will also use a `master` branch when built elsewhere, which is useful, because it prevents conflicts.

File: com.unity.UnityHub.yaml [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+base: io.atom.electron.BaseApp
+base-version: '18.08'
+runtime: org.freedesktop.Platform
+runtime-version: '18.08'
+sdk: org.freedesktop.Sdk
+command: start-unityhub
+separate-locales: false
+tags: [proprietary]
```
Comment (jurf) [inline com.unity.UnityHub.yaml]: No need to use JSON syntax here.

File: com.unity.UnityHub.yaml [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+finish-args:
+  - '--allow=multiarch'
```
Comment (jurf) [inline com.unity.UnityHub.yaml]: No need for quotes here, unless you prefer them. Same goes for the cleanups and build-commands, but really it’s your choice.

File: com.unity.UnityHub.yaml [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        sha256: 08a35e92deb3c85d269a0059a27d4140a9667a6369459299d08c17f713a92e73
+
+  - name: unityhub
+    buildsystem: simple
+    build-commands:
+      - 'install -Dm 755 unityhub.sh /app/bin/start-unityhub'
+      - 'install -Dm 755 vscode.py /app/bin/vscode-editor'
+      - 'install -Dm 755 apply_extra.sh /app/bin/apply_extra'
+      - 'install -Dm 755 patch-resources.py /app/bin/patch-resources'
+      - 'install -D com.unity.UnityHub.appdata.xml /app/share/appdata/com.unity.UnityHub.appdata.xml'
```
Comment (jurf) [inline com.unity.UnityHub.yaml]: `-Dm644`, same goes for the lower two lines.

Also, the AppData should go into `/app/share/metainfo`.

File: com.unity.UnityHub.appdata.xml (hunk 10 above)
Comment (refi64) [reply com.unity.UnityHub.appdata.xml]: @jurf The Hub itself is written in Electron 1, which uses GTK+2.

File: com.unity.UnityHub.yaml (hunk 18 above)
Comment (refi64) [reply com.unity.UnityHub.yaml]: I tend to prefer them, just because I know YAML can sometimes interpret things oddly (though flatpak-builder is relatively liberal with what it interprets as a string).

**Comment by refi64** [comment]
@ramcq @jurf Everything should be handled now; I commented some reasoning on the remaining two comments.

File: com.unity.UnityHub.yaml [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    version: '18.08'
+modules:
+  - name: compat
+    buildsystem: simple
+    build-commands:
+      - 'mkdir -p /app/lib/i386-linux-gnu /app/lib/debug/lib/i386-linux-gnu'
+
+  - shared-modules/glu/glu-9.0.0.json
+
+  - name: unappimage
```
Comment (barthalion) [inline com.unity.UnityHub.yaml]: AppImages can be self-extracted with `--appimage-extract`, why use additional tool?

Comment (barthalion) [reply com.unity.UnityHub.yaml]: OK, you actually explained it somewhere above in the discussion. Would be nice to have a comment in the manifest later.

**Comment by barthalion** [comment]
Repository has been created: https://github.com/flathub/com.unity.UnityHub
